### PR TITLE
feat: add ama-sdk create feature

### PR DESCRIPTION
**Pull request by kpanot**
## Proposed change

Add a feature to simplify the generation of an SDK repository

### CODE REVIEW

File: packages/@ama-sdk/create/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+const args = process.argv.slice(2);
+const argv = minimist(args);
+
+if (argv._.length === 0) {
+  // eslint-disable-next-line no-console
+  console.error('The project name is mandatory');
+  process.exit(-1);
+}
+
+let [name, pck] = argv._[0].replace(/^@/, '').split('/');
```
Comment (fpaul-1A) [inline packages/@ama-sdk/create/src/index.ts:18]: shouldn't this be `[pck, name]` ? or am I reading it wrong ?

Comment (fpaul-1A) [reply packages/@ama-sdk/create/src/index.ts:18]: nvm, `name` seems to be the npm package scope so it's fine, but it reads weird

**Review by fpaul-1A — APPROVED** [review]

File: packages/@ama-sdk/create/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+### NPM
+
+```shell
+npm create @ama-sdk <package-name> -- [--specPath=./path/to/spec.yaml]
+```
+
+### Yarn
+
+```shell
+yarn create @ama-sdk <project-name> [--specPath=./path/to/spec.yaml]
```
Comment (cpaulve-1A) [inline packages/@ama-sdk/create/README.md:21]: in the sdk generator schematics it is --spec-path, why the change?

File: packages/@ama-sdk/create/src/index.ts (hunk 1 above)
Comment (kpanot) [reply packages/@ama-sdk/create/src/index.ts:18]: I kept the naming of the generator options, I agree, it is weird :s

File: packages/@ama-sdk/create/README.md (hunk 2 above)
Comment (kpanot) [reply packages/@ama-sdk/create/README.md]: Wrong copy past in the doc, the parameter is actually `--spec-path`

File: packages/@ama-sdk/schematics/schematics/typescript/core/index.ts [hunk 3]
```diff
@@ -43,6 +43,7 @@ const getPathObjectTemplate = (pathObj: PathObject) => {
 export function ngGenerateTypescriptSDK(options: NgGenerateTypescriptSDKCoreSchematicsSchema): Rule {
 
   const specPath = path.resolve(process.cwd(), options.specPath);
+  const targetPath = options.directory ? options.directory : '/';
```
Comment (fpaul-1A) [inline packages/@ama-sdk/schematics/schematics/typescript/core/index.ts:46]:
```suggestion
  const targetPath = options.directory ?? '/';
```

File: packages/@ama-sdk/schematics/schematics/typescript/shell/index.ts [hunk 4]
```diff
@@ -61,12 +70,19 @@ export function ngGenerateTypescriptSDK(options: NgGenerateTypescriptSDKShellSch
       empty: ''
     };
 
+    const targetPath = options.directory ? options.directory : tree.root.path;
```
Comment (fpaul-1A) [inline packages/@ama-sdk/schematics/schematics/typescript/shell/index.ts:73]:
```suggestion
    const targetPath = options.directory ?? options.directory : tree.root.path;
```

**Review by fpaul-1A — DISMISSED** [review]

File: packages/@ama-sdk/schematics/README.md [hunk 5]
```diff
@@ -21,18 +23,17 @@ npx -p @angular/cli ng add @ama-sdk/core
 else
 
 ```shell
-# Install @angular-devkit/schematics-cli, @ama-sdk/core and the SDK generator
-yarn add --dev @angular-devkit/schematics-cli @ama-sdk/schematics @ama-sdk/core
-yarn schematics @ama-sdk/schematics:install
-yarn schematics @ama-sdk/core:install
+npm create @ama-sdk typescript <project-name> -- [--spec-path=./path/to/spec.yaml]
```
Comment (mrednic-1A) [inline packages/@ama-sdk/schematics/README.md:26]: `--` should be inside the brackets too 
`[-- --spec-path...}`

Comment (kpanot) [reply packages/@ama-sdk/schematics/README.md:26]: I followed Angular documentation (which is putting it outside)

**Review by fpaul-1A — DISMISSED** [review]

File: packages/@ama-sdk/schematics/schematics/typescript/core/index.ts [hunk 6]
```diff
@@ -107,19 +108,19 @@ export function ngGenerateTypescriptSDK(options: NgGenerateTypescriptSDKCoreSche
    */
   const updateSpec = (tree: Tree, _context: SchematicContext) => {
     const specContent = readFileSync(specPath).toString();
-    if (tree.exists('/readme.md')) {
+    if (tree.exists(path.join(targetPath, 'readme.md'))) {
```
Comment (mrednic-1A) [inline packages/@ama-sdk/schematics/schematics/typescript/core/index.ts:111]: `path.join(targetPath, 'readme.md')` could be extracted in a constant, as it is used in multiple places
same thing with `path.join(targetPath, 'swagger-spec.yaml')`

File: packages/@ama-sdk/schematics/schematics/typescript/create/index.ts [hunk 7]
```diff
@@ -6,6 +6,7 @@ import { NgGenerateTypescriptSDKCreateSchematicsSchema } from './schema';
 /**
  * Generate a typescript SDK repository and source code base on swagger specification
  *
+ * @deprecate use @ama-sdk/create package instead
```
Comment (mrednic-1A) [inline packages/@ama-sdk/schematics/schematics/typescript/create/index.ts:9]: I think it's this one we said we can remove. The package was not published before 9.x

File: packages/@ama-sdk/create/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+if (!pck) {
+  console.warn(`The given package name, is not a scoped package. a default "@${defaultScope}" scope will be used`);
+  pck = name;
+  name = defaultScope;
+}
+
+const targetDirectory = join('.', name, pck);
+const schematicsToRun = argv['spec-path'] ? ['@ama-sdk/schematics:typescript-shell', '@ama-sdk/schematics:typescript-core'] : ['@ama-sdk/schematics:typescript-shell'];
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/create/src/index.ts:34]: If there is another comment :)
```suggestion
const schematicsToRun = [
   '@ama-sdk/schematics:typescript-shell',
   ...(argv['spec-path'] ?  ['@ama-sdk/schematics:typescript-core'] : [])
 ];
```

File: packages/@ama-sdk/schematics/schematics/code-generator/open-api-cli-generator/open-api-cli.generator.ts [hunk 9]
```diff
@@ -1,14 +1,25 @@
...
+    return this.packageManager === 'npm' ? 'npx' : 'yarn';
+  }
+
   /** @inheritDoc */
-  protected generatorName = 'openapi-codegen-npm-cli';
+  protected readonly generatorName = 'openapi-codegen-npm-cli';
+
+  constructor(options?: { packageManager?: 'npm' | 'yarn' | '' }) {
+    super();
+    this.packageManager = options?.packageManager || (process.env?.npm_execpath?.includes('yarn') ? 'yarn' : 'npm');
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/schematics/schematics/code-generator/open-api-cli-generator/open-api-cli.generator.ts:21]: can't we do a global helper for that ?
we have this kind of piece of code everywhere

File: packages/@ama-sdk/schematics/schematics/code-generator/open-api-cli-generator/open-api-cli.generator.ts [hunk 10]
```diff
@@ -65,7 +75,9 @@ export class OpenApiCliGenerator extends CodeGenerator<OpenApiCliOptions> {
 
   /** @inheritDoc */
   protected runCodeGeneratorFactory = (factoryOptions: { rootDirectory?: string } = {}) => {
-    const rootDirectory = factoryOptions.rootDirectory || process.cwd();
+    const rootDirectory = factoryOptions.rootDirectory ?
+      (path.isAbsolute(factoryOptions.rootDirectory) ? factoryOptions.rootDirectory : path.resolve(process.cwd(), factoryOptions.rootDirectory)) :
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/schematics/schematics/code-generator/open-api-cli-generator/open-api-cli.generator.ts:79]: `path.resolve` or `path.posix.resolve` ?
When should we use `posix` and when not ?

File: packages/@ama-sdk/schematics/schematics/index.it.spec.ts [hunk 11]
```diff
@@ -72,6 +72,8 @@ function setupYarn(yarnVersion: string) {
 
 /**
  * Setup a new sdk using schematics CLI
+ *
+ * @deprecated A new SDK should be generated via the `npm create` feature
```
Comment (matthieu-crouzet) [inline packages/@ama-sdk/schematics/schematics/index.it.spec.ts:76]: you are in the IT test why not using the `npm create` in this one ?

File: packages/@ama-sdk/schematics/README.md (hunk 5 above)
Comment (matthieu-crouzet) [inline packages/@ama-sdk/schematics/README.md:26]: In fact the `(npm|yarn) create` is developed to create a repo from scratch not inside an otter project 
maybe it could have a dedicated section

File: packages/@ama-sdk/create/src/index.ts (hunk 8 above)
Comment (kpanot) [reply packages/@ama-sdk/create/src/index.ts:34]: Hmmm why not, let's take it :D

File: packages/@ama-sdk/schematics/schematics/index.it.spec.ts (hunk 11 above)
Comment (kpanot) [reply packages/@ama-sdk/schematics/schematics/index.it.spec.ts:76]: Because this is not the `@ama-sdk/create` package and using it here will create a circular dep.
The IT of the `@ama-sdk/create` will be created in the dedicated package but this function is actually deprecated and need to be removed (which will lead to the rewrite of these ITs which is not the purpose of this PR)

File: packages/@ama-sdk/schematics/schematics/code-generator/open-api-cli-generator/open-api-cli.generator.ts (hunk 10 above)
Comment (fpaul-1A) [reply packages/@ama-sdk/schematics/schematics/code-generator/open-api-cli-generator/open-api-cli.generator.ts:80]: my understanding is that we need to use `posix` at least when building paths for the Tree of schematics and for globby
when targeting an actual on the disk we should use `path.resolve` to match the OS paths

also, for the tree of schematics we should not use `resolve` as it create an absolute path

Comment (kpanot) [reply packages/@ama-sdk/schematics/schematics/code-generator/open-api-cli-generator/open-api-cli.generator.ts:80]: This is for the `cwd` option of `spawn` function, so it should be aligned with OS path separator.
Globally par default it is **not** `posix` except when we want a path to print in a `json` file (cd tsconfig.json, angular.json, etc...) or use it in `tree` or `require` paths

**Review by fpaul-1A — APPROVED** [review]

**Review by fpaul-1A — APPROVED** [review]
